src/pages/developers/tutorials/localnet.mdx

```diff
@@ -0,0 +1,173 @@
+```
+
+The EVM gateway executes `onRevert` function of the Receiver contract:
+
+```
+[EVM]: Contract 0x9A676e781A523b5d0C0e43731313A708CB607508 executing onRevert (context: {"asset":"0x0000000000000000000000000000000000000000","amount":0,"revertMessage":"0x3078"})
+```
+
+## Call a Receiver Contract from a Universal App
+
+Execute the `callFromZetaChain` function of the universal app:
+
+```
+yarn hardhat call-from-zetachain --message bob --network localhost --revert-address 0x9A676e781A523b5d0C0e43731313A708CB607508 --revert-message "my revert message" --call-on-revert --contract 0x0DCd1Bf9A1b36cE34237eEaFef220932846BCD82 --receiver 0x9A676e781A523b5d0C0e43731313A708CB607508
+```
+
+The function calls the ZetaChain gateway contract. The ZetaChain gateway emits a
+"Called" event acknowledging that the call has been successful.
+
+```
+[ZetaChain]: Gateway: 'Called' event emitted
+```
+
+Localnet picks up the "Called" event and calls the EVM gateway `execute`
+function, which calls the Receiver contract:
+
+```
+[EVM]: Calling 0x9a676e781a523b5d0c0e43731313a708cb607508 with message 0xa777d0dc00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003626f620000000000000000000000000000000000000000000000000000000000
+```
+
+The Receiver contract then emits an event:
+
+```
+[EVM]: Event from contract: {"_type":"log","address":"0x9A676e781A523b5d0C0e43731313A708CB607508","blockHash":"0x0deb4a3e22b73320ffad3d916c5e2b2db450dddc162feaafa84546612cebd238","blockNumber":22,"data":"0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003626f620000000000000000000000000000000000000000000000000000000000","index":0,"removed":false,"topics":["0xfa051318aca3e493e16e6cf4bccc017c7e061924a42ef27b3bb373c4707b636a"],"transactionHash":"0x94c4d53aed68c97f18e467207ff6cdc1e855c65bcddab399f100fe2786ee33a8","transactionIndex":0}
+```
```

Comment:
I think a comparison between localnet and the real networks would be helpful to close this out. 

Something like "Interacting between real networks like ZetaChain and Ethereum follows the same steps as Localnet. Developers or users can interact with the EVM gateway on live networks like Ethereum, BSC, or Base and call contracts on ZetaChain or vice versa."

Reply:
1cf9c57